packages/opentelemetry-context-base/src/types.ts

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export interface BaseContext {
```

Comment:
I still think a simpler approach that doesn't support getting and setting individual properties would serve us better internally. Then we can always extend with a separate package if needed.

Something like:

```ts
export interface BaseContext {
  active: unknown; // the current active context
  activate<T>(object: unknown, fn: T); // activate an object on the current context for the given function
  bind<T>(target: T, object?: unknown): T; // "wrap" is semantically incorrect for objects
  enable(): void; // could also return the instance
  disable(): void;
}
```

I also think using "scope" instead of "context", similar to OpenTracing, would be better given how "context" is already overloaded in so many ways in the project already.

Reply:
A couple questions on this:
- Would it make sense to have `object` typed as `{[key: string]: unknown}` so that callers could depend on the context object being a map? That could allow us to e.g. store some OpenTelemetry specific context in a given key on it without disrupting other users of the context.
- Could you help me understand the `activate` function? I assume that `bind` creates a new function with hooks around the passed function such that when the new bound function it will invoke `target`  (or the emitted event handler, etc.) such that its current will be set to the current or specified context. Is `activate` saying that it will change the default context for the function in a mutating way? Does it call the function right away?

Reply:
> That could allow us to e.g. store some OpenTelemetry specific context in a given key on it without disrupting other users of the context.

Other users of would use their own `BaseContext` instance so there would not be a conflict.

> Could you help me understand the activate function?

The main difference is `activate()` is meant to just activate a span immediately in the callnack, but `bind()` returns a wrapper that when called will pass its argument to the delegate and activate the span as well.

For example:

```js
scope.activate(span, () => {
  // no arguments are passed
  // scope.active() returns the span in every async context created here
})

const fn = (foo) => {
  // foo='bar'
  // scope.active() returns the span from below in every async context created here
}

const wrapper = scope.bind(fn, span)

wrapper('bar')
```

Reply:
OK, I see! I think I'm used to that function being called something like `run` or some derivative e.g. `runInAsyncScope`.

Reply:
I find the term `activate` on the object not really transparent to it's behavior, what do you think about `setAsCurrent` and `current()` or `setAsActive` and `active` ?

Reply:
Also, i agree with @rochdev that the `context` word is already used a lot in the implementation and using the `scope` word is not wrong, so i'm in favor of using it

Reply:
You are not really setting it as current though. You are activating the span on the scope of the function that is passed as a parameter. I'm not sure if it's 100% clear, but it's definitely semantically correct.

Zones use the work `fork` for this. CLS uses `run`. The spec uses `withSpan` but on the tracer.

My personal preference would be to find any single word that correctly represents what this does, which I think `activate` does but I may be biased as I've been using this terminology for a long time from [OpenTracing](https://github.com/opentracing/specification/blob/master/rfc/scope_manager.md#scopemanager) at this point.

Reply:
I would personally prefer to use `with` since it's in the spec and clearer than `active` to me

Reply:
So `with()` and `current()`? Do you think `bind()` makes sense as well?

Reply:
Totally fine with `bind`